crypto/x509/by_store.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/store.h>
+#include "internal/cryptlib.h"
+#include "internal/x509_int.h"
+#include "x509_lcl.h"
+
+/* Generic object loader, given expected type and criterion */
+static int cache_objects(X509_LOOKUP *lctx, const char *uri,
+                         X509_LOOKUP_TYPE type,
+                         const OSSL_STORE_SEARCH *criterion,
+                         int depth)
+{
+    int ok = 0;
+    OSSL_STORE_CTX *ctx = NULL;
+    int expect = 0;
+    X509_STORE *xstore = X509_LOOKUP_get_store(lctx);
+
+    if (type == X509_LU_X509)
+        expect = OSSL_STORE_INFO_CERT;
+    else if (type == X509_LU_X509)
```

Comment:
Doubling lookup type.  Second one is for X509_LU_CRL.

Please note that cache_objects() is usable in context on dynamic load(="add") when lookup method by_subject is invoked.
It you decide to use in general (static "load") cache_objects() has to support "none" lookup.

Samples:
(a) "static" load (without error checks):
		lookup = X509_STORE_add_lookup(x509store, X509_LOOKUP_store());
		iX509_LOOKUP_load_store(lookup, url);
In this case is expected immediately all for external url to be loaded either cert or crl.

(b) dynamic (on request):
		lookup = X509_STORE_add_lookup(x509store, X509_LOOKUP_store());
		X509_LOOKUP_add_store(lookup, _locations->ldap_url);
In this case load is postponed to by_subject and at time of invocation we know lookup "type".

I prefer only "add" (dynamic) functionality for "store" lookup.